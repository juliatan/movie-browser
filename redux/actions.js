import { fetchMovies } from '../api';

// action types -> done to prevent typos and for easier bug identification
export const SEARCH_TERM_SENT = 'SEARCH_TERM_SENT'
export const NEW_RESULTS_RECEIVED = 'NEW_RESULTS_RECEIVED'
export const ADDITIONAL_RESULTS_RECEIVED = 'ADDITIONAL_RESULTS_RECEIVED'
export const RESULTS_ERROR = 'RESULTS_ERROR'
export const CLEAR_RESULTS = 'CLEAR_RESULTS'

// async action creator

// invoked directly from SearchScreen.js. This dispatches various different "types" to reducer.js.
// Because dispatch is a function, we need to use 3rd party "redux-thunk" library and import "applyMiddleware" from redux in store.js.
export const returnResults = (searchTerm, pageNumber) => async dispatch => {
  dispatch({type: SEARCH_TERM_SENT})
  try {
    const response = await fetchMovies(searchTerm, pageNumber)
    const results = response.results

    const totalResults = response.totalResults
    const maxPages = Math.ceil(totalResults/10) // 10 results per page

    if (pageNumber === 1) {
      dispatch({type: NEW_RESULTS_RECEIVED, payload: { movies: results, maxPages: maxPages }})
    } else {
      dispatch({type: ADDITIONAL_RESULTS_RECEIVED, payload: { movies: results }})
    }

  } catch (err) {
    console.error('API error')
    dispatch({type: RESULTS_ERROR, payload: 'API error'})
  }
}

export const clearResults = () => dispatch => {
  dispatch({type: CLEAR_RESULTS})
}