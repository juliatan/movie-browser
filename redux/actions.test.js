import * as actions from './actions'

test('clearResults returns an action', () => {
  expect(actions.clearResults()).toEqual({
    type: actions.CLEAR_RESULTS
  })
})

describe('returnResults returns actions', () => {
  const fakeFetchedObject = {results: [], totalResults: 20}
  const fakeResultsObject1 = {movies: [], maxPages: 2}
  const fakeResultsObject2 = {movies: []}
  const errMessage = 'API error'
  const mockFetch = (searchTerm, pageNumber) => {
    if (searchTerm === 'test' && pageNumber === 1) {
      return fakeFetchedObject
    }

    if (searchTerm === 'test' && pageNumber === 2) {
      return fakeFetchedObject
    }

    throw new Error(errMessage)
  }

  it('dispatches SEARCH_TERM_SENT', async () => {
    const mockDispatch = jest.fn()
    await actions.returnResults('', '')(mockDispatch)
      // mockDispatch.mock.calls all the args that the mock fn was invoked on
    expect(mockDispatch.mock.calls[0][0]).toEqual({type: actions.SEARCH_TERM_SENT})
  })

  it('dispatches NEW_RESULTS_RECEIVED with search term and page 1', async () => {
    const mockDispatch = jest.fn()
    await actions.returnResults('test', 1, mockFetch)(mockDispatch)
    expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.NEW_RESULTS_RECEIVED, payload: fakeResultsObject1})
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
  })

  it('dispatches ADDITIONAL_RESULTS_RECEIVED with search term and page 2', async () => {
    const mockDispatch = jest.fn()
    await actions.returnResults('test', 2, mockFetch)(mockDispatch)

    expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.ADDITIONAL_RESULTS_RECEIVED, payload: fakeResultsObject2})
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
  })

  it('dispatches RESULTS_ERROR with incorrect arguments', async () => {
    const mockDispatch = jest.fn()
    await actions.returnResults(test, 1000, mockFetch)(mockDispatch)

    expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.RESULTS_ERROR, payload: errMessage})
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
  })
})