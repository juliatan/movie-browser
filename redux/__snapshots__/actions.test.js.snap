// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returnResults returns actions dispatches ADDITIONAL_RESULTS_RECEIVED with search term and page 2 1`] = `
Array [
  Object {
    "payload": Object {
      "movies": Array [],
    },
    "type": "ADDITIONAL_RESULTS_RECEIVED",
  },
]
`;

exports[`returnResults returns actions dispatches NEW_RESULTS_RECEIVED with search term and page 1 1`] = `
Array [
  Object {
    "payload": Object {
      "maxPages": 2,
      "movies": Array [],
    },
    "type": "NEW_RESULTS_RECEIVED",
  },
]
`;

exports[`returnResults returns actions dispatches RESULTS_ERROR with incorrect arguments 1`] = `
Array [
  Object {
    "payload": "API error",
    "type": "RESULTS_ERROR",
  },
]
`;
